### PR TITLE
plugins/comment-box-nvim: init

### PR DESCRIPTION
**Pull request by elythh**
Adds comment-box-nvim as a plugin.

Waiting on https://nixpk.gs/pr-tracker.html?pr=335126

### CODE REVIEW

File: plugins/utils/comment-box.nix [hunk 1]
```diff
@@ -0,0 +1,149 @@
+{
+  lib,
+  helpers,
+  config,
+  pkgs,
+  ...
+}:
+with lib;
```
Comment (khaneliman) [inline plugins/utils/comment-box.nix:8]:
```suggestion
let
inherit (lib) types;
in
```

File: plugins/utils/comment-box.nix [hunk 2]
```diff
@@ -0,0 +1,149 @@
+{
+  lib,
+  helpers,
```
Comment (khaneliman) [inline plugins/utils/comment-box.nix:3]:
```suggestion
```

File: plugins/utils/comment-box.nix [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+  config,
+  pkgs,
+  ...
+}:
+with lib;
+helpers.neovim-plugin.mkNeovimPlugin config {
```
Comment (khaneliman) [inline plugins/utils/comment-box.nix:9]:
```suggestion
lib.nixvim.neovim-plugin.mkNeovimPlugin config {
```

File: plugins/utils/comment-box.nix [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+  name = "comment-box";
+  originalName = "comment-box.nvim";
+  luaName = "comment-box";
+  defaultPackage = pkgs.vimPlugins.comment-box-nvim;
+
+  maintainers = [ maintainers.elythh ];
```
Comment (khaneliman) [inline plugins/utils/comment-box.nix:15]:
```suggestion
  maintainers = [ lib.maintainers.elythh ];
```

File: plugins/utils/comment-box.nix [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+  settingsOptions = with helpers.defaultNullOpts; {
```
Comment (khaneliman) [inline plugins/utils/comment-box.nix:17]:
```suggestion
  settingsOptions = with lib.nixvim.defaultNullOpts; {
```

File: plugins/utils/comment-box.nix [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    comment_style =
+      helpers.defaultNullOpts.mkEnum
```
Comment (khaneliman) [inline plugins/utils/comment-box.nix:19]:
```suggestion
      lib.nixvim.defaultNullOpts.mkEnum
```

**Comment by elythh** [comment]
waiting on https://nixpk.gs/pr-tracker.html?pr=335126

File: plugins/utils/comment-box.nix [hunk 7]
```diff
@@ -0,0 +1,150 @@
+{
+  lib,
+  config,
+  pkgs,
+  ...
+}:
+let
+  inherit (lib) types;
```
Comment (khaneliman) [inline plugins/utils/comment-box.nix:8]:
```suggestion
  inherit (lib) mkAttrsOf mkBool mkInt mkStr types;
```
Might need to be something like this for all the others that were from the `with lib`

**Comment by khaneliman** [comment]
Also need some tests added for the new plugin. Example of a recent new plugin https://github.com/nix-community/nixvim/pull/2000/files#diff-30f50cb437e6e64c5beeebeee45f2c3d0968b55005e47f457345c83ba941ddfb

File: plugins/utils/comment-box.nix (hunk 7 above)
Comment (khaneliman) [reply plugins/utils/comment-box.nix]: Sorry, I didn't see you used the inherit from defaultNullOpts in a with block.. you can revert that last one.

File: plugins/utils/comment-box.nix [hunk 8]
```diff
@@ -0,0 +1,160 @@
+{
+  lib,
+  config,
+  pkgs,
+  ...
+}:
+let
+  inherit (lib)
```
Comment (khaneliman) [inline plugins/utils/comment-box.nix:8]:
```suggestion
  inherit (lib) types;
  inherit (lib.nixvim) defaultNullOpts;
```

File: plugins/utils/comment-box.nix [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+          "line"
+          "block"
+          "auto"
+        ]
+        "line"
+        ''
+          Select the type of comments. Default: 'line'.
+        '';
+
+    doc_width = mkInt types.number 80 ''
```
Comment (khaneliman) [inline plugins/utils/comment-box.nix:41]: We should make it more clear where these are coming from to fit in line with our other usage across codebase.This will apply to all options, just showing suggestion on this one to avoid repetition. 

```suggestion
    doc_width = defaultNullOpts.mkInt types.number 80 ''
```

File: plugins/utils/comment-box.nix [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+      Width of the document. Default: 80.
+    '';
+    box_width = mkInt types.number 60 ''
+      Width of the boxes. Default: 60.
+    '';
+
+    borders = mkAttrsOf types.srt {
```
Comment (khaneliman) [inline plugins/utils/comment-box.nix:48]:
```suggestion
    borders = defaultNullOpts.mkAttrsOf types.str {
```

File: plugins/utils/comment-box.nix (hunk 7 above)
Comment (elythh) [reply plugins/utils/comment-box.nix:8]: No worries, both are fine to me. I'll keep it that way since I find it to be easier for the eyes too

File: plugins/utils/comment-box.nix (hunk 9 above)
Comment (elythh) [reply plugins/utils/comment-box.nix]: Done https://github.com/nix-community/nixvim/compare/df79c0b795e86a4af80c342037bc4fd31dda95e0..a1761087fa9bbaef4cf95db6102f2c88ba48893b

File: plugins/utils/comment-box.nix [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+    comment_style =
+      defaultNullOpts.mkEnum
+        [
+          "line"
+          "block"
+          "auto"
+        ]
+        "line"
+        ''
+          Select the type of comments. Default: 'line'.
```
Comment (khaneliman) [inline plugins/utils/comment-box.nix:31]:
```suggestion
          Select the type of comments.
```
The default will show up in the UI because of the "line" pluginDefault declared above. We can remove all these `Default:`

Comment (khaneliman) [reply plugins/utils/comment-box.nix:31]: You can see what I'm talking about if you temporarily change the defaultPlugin to something that will build and run `serve-docs` command from the devShell.

File: plugins/utils/comment-box.nix [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+
+      line_blank_line_below = defaultNullOpts.mkBool false ''
+        Insert a blank line below the line.
+      '';
+    };
+  };
+
+  settingsExample = {
+    comment_style = "line";
+    doc_width = 80; # width of the document
```
Comment (khaneliman) [inline plugins/utils/comment-box.nix:125]: We should probably remove all these comments, too. They won't be rendered in the page that gets created and they look like repeats of the descriptions for the options above that will.

Comment (elythh) [reply plugins/utils/comment-box.nix:125]: oh yes, I thought the raw nix code was shown in the example

Comment (elythh) [reply plugins/utils/comment-box.nix]: https://github.com/nix-community/nixvim/compare/3dac1a6e16d53c29d540cf2dc3405265cac1df66..c4c237e55eac9e5bdeb41edfcc004c2bca5b876a

**Comment by elythh** [comment]
thanks for the review @khaneliman !

File: plugins/utils/comment-box.nix [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+        Insert a blank line above the line.
+      '';
+
+      line_blank_line_below = defaultNullOpts.mkBool false ''
+        Insert a blank line below the line.
+      '';
+    };
+  };
+
+  settingsExample = {
```
Comment (khaneliman) [inline plugins/utils/comment-box.nix:123]: We also want this section to only be non defaults to help people see an example configuration that changes behavior. Do you have a personal config that you plan to use or can just tweak the values to be opposite of defaults for booleans and maybe increase widths.

**Comment by khaneliman** [comment]
LGTM, thanks for your addition! Plugin actually looks very appealing, too. I'm definitely going to use this as soon as its available.

File: plugins/utils/comment-box.nix (hunk 13 above)
Comment (elythh) [reply plugins/utils/comment-box.nix:123]: Updated with what would a random configuration look like, only including the modified options.

File: plugins/utils/comment-box.nix (hunk 12 above)
Comment (MattSturgeon) [reply plugins/utils/comment-box.nix]: If you _want_ raw nix code to show up in the example, you can use `lib.literalExpression` (or `lib.literalMD` for injecting markdown).

e.g.

```nix
settingsExample = lib.literalExpression ''
  {
    comment_style = "line";
    doc_width = 80; # width of the document
    # etc
  }
'';

Most of the time we prefer just passing a nix value instead of a literalExpression string, since it ensures a somewhat standardized formatting, and ensures the syntax is valid, but I've nothing against using `literalExpression` where it makes sense to do so.

**Comment by elythh** [comment]
looks like it's been upstreamed in nixpkgs-unstable, the input needs to be updated tho
cc @khaneliman @MattSturgeon

**Comment by khaneliman** [comment]
Created https://github.com/nix-community/nixvim/pull/2090 to update flake.lock.

**Comment by khaneliman** [comment]
Should be able to rebase your branch, now.

File: plugins/utils/comment-box.nix [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+
+      line_end = defaultNullOpts.mkStr "─" ''
+        Symbol used to draw the end of a line.
+      '';
+
+      title_left = defaultNullOpts.mkStr "─" ''
+        Symbol used to draw the left border of the title.
+      '';
+
+      tttle_right = defaultNullOpts.mkStr "─" ''
```
Comment (khaneliman) [inline plugins/utils/comment-box.nix:97]:
```suggestion
      title_right = defaultNullOpts.mkStr "─" ''
```

File: plugins/utils/comment-box.nix [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+  pkgs,
+  ...
+}:
+let
+  inherit (lib.nixvim) defaultNullOpts;
+in
+lib.nixvim.neovim-plugin.mkNeovimPlugin config {
+  name = "comment-box";
+  originalName = "comment-box.nvim";
+  luaName = "comment-box";
```
Comment (khaneliman) [inline plugins/utils/comment-box.nix:13]:
```suggestion
```

Not needed when it matches the name of the plugin. But, not a big deal.

File: plugins/utils/comment-box.nix (hunk 14 above)
Comment (elythh) [reply plugins/utils/comment-box.nix:97]: nice catch

**Review by khaneliman — APPROVED** [review]
LGTM, thanks for the addition!

**Comment by khaneliman** [comment]
@mergifyio queue

File: lib/maintainers.nix [hunk 16]
```diff
@@ -20,6 +20,12 @@
     name = "Daniel Laing";
     keys = [ { fingerprint = "0821 8B96 DC73 85E5 BB7C  A535 D264 3BD2 13BC 0FA8"; } ];
   };
+  elythh = {
+    email = "gwen@omg.lol";
+    github = "elythh";
+    githubId = 50964650;
+    name = "gwen";
+  };
```
Comment (MattSturgeon) [inline lib/maintainers.nix:28]: FYI @khaneliman: ideally we like new maintainers to be in their own commit, separate from the new plugin.

Comment (khaneliman) [reply lib/maintainers.nix:28]: Ugh don’t know how I missed that

Comment (MattSturgeon) [reply lib/maintainers.nix:28]: Easily done. Commits + commit messages are hidden out of the way when reviewing on GitHub, most of the time we just focus on the overall diff.
